raw-spans-grouper/src/main/java/org/hypertrace/core/rawspansgrouper/RawSpansGroupingTransformer.java

```diff
@@ -0,0 +1,132 @@
+package org.hypertrace.core.rawspansgrouper;
+
+import static org.hypertrace.core.rawspansgrouper.RawSpanGrouperConstants.INFLIGHT_TRACE_STORE;
+import static org.hypertrace.core.rawspansgrouper.RawSpanGrouperConstants.OUTPUT_TOPIC_PRODUCER;
+import static org.hypertrace.core.rawspansgrouper.RawSpanGrouperConstants.RAW_SPANS_GROUPER_JOB_CONFIG;
+import static org.hypertrace.core.rawspansgrouper.RawSpanGrouperConstants.SPAN_GROUPBY_SESSION_WINDOW_INTERVAL_CONFIG_KEY;
+import static org.hypertrace.core.rawspansgrouper.RawSpanGrouperConstants.TRACE_EMIT_TRIGGER_STORE;
+
+import com.typesafe.config.Config;
+import java.time.Duration;
+import java.time.Instant;
+import org.apache.kafka.streams.KeyValue;
+import org.apache.kafka.streams.kstream.Transformer;
+import org.apache.kafka.streams.processor.Cancellable;
+import org.apache.kafka.streams.processor.ProcessorContext;
+import org.apache.kafka.streams.processor.PunctuationType;
+import org.apache.kafka.streams.processor.To;
+import org.apache.kafka.streams.state.KeyValueIterator;
+import org.apache.kafka.streams.state.KeyValueStore;
+import org.apache.kafka.streams.state.ValueAndTimestamp;
+import org.hypertrace.core.datamodel.RawSpan;
+import org.hypertrace.core.datamodel.RawSpans;
+import org.hypertrace.core.datamodel.StructuredTrace;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Receives spans keyed by trace_id and stores them. A {@link TraceEmitPunctuator} is scheduled to
+ * run after the {@link RawSpansGroupingTransformer#sessionTimeoutMs} interval to emit the trace. If
+ * any spans for the trace arrive within the {@link RawSpansGroupingTransformer#sessionTimeoutMs}
+ * interval then the {@link RawSpansGroupingTransformer#sessionTimeoutMs} will get reset and the
+ * trace will get an additional {@link RawSpansGroupingTransformer#sessionTimeoutMs} time to accept
+ * spans.
+ */
+public class RawSpansGroupingTransformer implements
+    Transformer<String, RawSpan, KeyValue<String, StructuredTrace>> {
+
+  private static final Logger logger = LoggerFactory.getLogger(RawSpansGroupingTransformer.class);
+  private ProcessorContext context;
+  private KeyValueStore<String, ValueAndTimestamp<RawSpans>> inflightTraceStore;
+  private KeyValueStore<String, Long> traceEmitTriggerStore;
+  private long sessionTimeoutMs;
+  private To outputTopic;
+
+  @Override
+  public void init(ProcessorContext context) {
+    this.context = context;
+    this.inflightTraceStore = (KeyValueStore<String, ValueAndTimestamp<RawSpans>>) context
+        .getStateStore(INFLIGHT_TRACE_STORE);
+    this.traceEmitTriggerStore = (KeyValueStore<String, Long>) context
+        .getStateStore(TRACE_EMIT_TRIGGER_STORE);
+    this.sessionTimeoutMs = ((Config) (context.appConfigs().get(RAW_SPANS_GROUPER_JOB_CONFIG)))
```

Comment:
do we want to use `sessionTimeoutMs` -> `groupingWindowTimeoutMs`?

Reply:
done